Play internet radios in MediaElement

fxthomas opened the pull request:
This is an attempt at fixing #408, which transfers playback of internet radios to a MediaElement player.

Remaining issues:
- [x] Make internet radios play in a MediaElement player
- [x] Correctly handle switching between media file playback ←→ radio playback
- [x] Keep previously playing radio when the page is reloaded
- [x] Correctly disable unavailable UI elements when playing radios
- [x] Show parsing error notifications in the UI instead of in the logs
- [x] Add a reasonable limit (e.g. 50) on the number of source URLs fetched from the radio server
- [x] ~~Check validity of radio URL when adding in the Settings page~~
- [x] Find out how to handle recursive playlists (do we handle them at all?)

Testing:
- [x] Test with other types of players (cast, jukebox)
- [x] Test with many types of internet radios
- [x] Test cases where the playlist cannot be parsed

fxthomas commented:
The main feature should work now. Please test and review whenever you can, I'll also load this on my server and keep testing it!

fxthomas commented:
I especially welcome comments/tests about how I handled player modes other than the default (e.g. Jukebox mode), I don't use these often enough to be 100% sure of how they'll behave with internet radios.

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueInfo.java:
```diff
@@ -74,10 +76,12 @@ public boolean isRepeatEnabled() {
+        return shuffleRadioEnabled;
     }
 
+    public boolean isInternetRadioEnabled() { return internetRadioEnabled; }
```
Please put this in a new line the other ones :)

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueService.java:
```diff
@@ -232,6 +242,20 @@ public PlayQueueInfo play(int id) throws Exception {
+     */
+    public PlayQueueInfo playInternetRadio(int id, Integer index) throws Exception {
+        HttpServletRequest request = WebContextFactory.get().getHttpServletRequest();
+        HttpServletResponse response = WebContextFactory.get().getHttpServletResponse();
```
Why do you need the response?

fxthomas replied:
Ah, you're right, that's a copy-paste from other methods. Removed.

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueService.java:
```diff
@@ -645,20 +685,44 @@ private PlayQueueInfo convert(HttpServletRequest request, Player player, boolean
-        Locale locale = RequestContextUtils.getLocale(request);
 
-        List<PlayQueueInfo.Entry> entries = new ArrayList<PlayQueueInfo.Entry>();
+        float gain = 0.0f;
```
This could be on a single line

fxthomas replied:
Changed.

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueService.java:
```diff
@@ -767,9 +872,7 @@ public void setJukeboxService(JukeboxService jukeboxService) {
-    public void setTranscodingService(TranscodingService transcodingService) {
-        this.transcodingService = transcodingService;
-    }
+    public void setTranscodingService(TranscodingService transcodingService) { this.transcodingService = transcodingService; }
```
Why did you put it on a single line?

fxthomas replied:
Changed.

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueService.java:
```diff
@@ -803,7 +906,7 @@ public void setPlaylistService(PlaylistService playlistService) {
-    public void setJwtSecurityService(JWTSecurityService jwtSecurityService) {
-        this.jwtSecurityService = jwtSecurityService;
-    }
+    public void setJwtSecurityService(JWTSecurityService jwtSecurityService) { this.jwtSecurityService = jwtSecurityService; }
```
Why on a single line?

fxthomas replied:
Changed.

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/domain/InternetRadioSource.java:
```diff
@@ -0,0 +1,11 @@
+package org.airsonic.player.domain;
+
+public class InternetRadioSource {
```
Is it really worth having a class to wrap a String?

fxthomas replied:
Maybe that's just personal preference, but what the code does sounds clearer to me when it's handling a "more qualified" type like `List<InternetRadioSource>` instead of a regular `List<String>`. Also, it's not a particularly performance-sensitive section, otherwise I'd agree with you.

jvoisin replied:
Too bad there isn't a nice way to do type aliasing in Java :/

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/service/InternetRadioService.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    private List<InternetRadioSource> retrieveInternetRadioSources(InternetRadio radio) throws Exception {
+
```
No need for a new line here

fxthomas replied:
Yup ;)

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/service/InternetRadioService.java:
```diff
@@ -0,0 +1,112 @@
+            inputPlaylist = SpecificPlaylistFactory.getInstance().readFrom(playlistUrl);
+        } catch (Exception e) {
+            LOG.error("Unable to parse internet radio playlist: {}", playlistUrl.toString(), e);
+            throw e;
```
Don't we want to simply return an empty list, instead of yielding the exception?

fxthomas replied:
Hmm, I haven't managed yet to work on forwarding parsing errors to the UI, which is why I defaulted to throwing the exception. Any ideas on how to do this properly?

jvoisin replied:
I guess that we should add an helper to do this then, in an other PR.
I guess yielding an exception is fine for now :)

jvoisin commented on airsonic-main/src/main/webapp/WEB-INF/jsp/playQueue.jsp:
```diff
@@ -447,11 +469,24 @@
+
+            if (!internetRadioEnabled) {
+                // Show star/remove buttons in all cases...
+                $("#starSong" + id).css("display", "");
```
What about something like `$("#starSong" + id).show();` instead?

fxthomas replied:
Yup, it does work better! Changed.

jvoisin commented on airsonic-main/src/main/webapp/WEB-INF/jsp/playQueue.jsp:
```diff
@@ -447,11 +469,24 @@
             } else {
-                $("#starSong" + id).attr("src", "<spring:theme code='ratingOffImage'/>");
+                // ...except from when internet radio is playing.
+                $("#starSong" + id).css("display", "none");
```
What about something like `$("#starSong" + id).hide();` instead?

fxthomas replied:
Same, changed.

fxthomas commented:
Thanks for the comments, I rebased and fixed a few things you mentioned. I also added code to hide some useless UI items (such as playlist management buttons) when playing a radio.

fxthomas commented:
After running a few tests, some radios (e.g. https://www.internet-radio.com/station/edmpulse/) add redirect URLs in the .m3u files (M3U in M3U, sometimes on multiple levels). How do you think these should handled?

jvoisin commented:
I don't know much about webradio nor their usages, so no idea beside following redirections when it makes sense I guess?

fxthomas commented:
FWIW DSub doesn't handle the URLs I used for testing either, so I guess we did not lose any functionality.

So, HTTP redirects should work fine, but playlists-in-playlists will not work for now.

fxthomas commented:
* Added basic error handling, using the "toast" system already in place. In case an error happens, the full error is logged internally but the radio service returns an empty list.
* Added some conservative limits (100KB, 250 URLs) on what we parse from remote sources, in order to avoid trying to load a full response from a remote server in case the playlist data is mistakenly (or maliciously) large.

fxthomas commented:
(An interesting note I had is that it should be possible to link playlists from two Airsonic instances together. Airsonic federation could be a thing!)

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueService.java:
```diff
@@ -645,20 +684,43 @@ private PlayQueueInfo convert(HttpServletRequest request, Player player, boolean
+        Locale locale = RequestContextUtils.getLocale(request);
         PlayQueue playQueue = player.getPlayQueue();
 
+        List<PlayQueueInfo.Entry> entries = new ArrayList<PlayQueueInfo.Entry>();
```
No need to specify the type on the right side, using the diamond operator (`<>`) is the recommended way in *modern* Java.

fxthomas replied:
Good to know, thanks!

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueService.java:
```diff
@@ -677,12 +739,53 @@ private PlayQueueInfo convert(HttpServletRequest request, Player player, boolean
+            entries.add(new PlayQueueInfo.Entry(
+                    streamId,          // Entry id
+                    streamTrackNumber, // Track number
+                    streamTitle,       // Use URL as stream title
```
Since there is a comment here, I'm not convinced that it's worth having a temporary variable, is it?

fxthomas replied:
Yup, you're right. Tell me if the new commit sounds okay with you.

jvoisin commented on airsonic-main/src/main/webapp/WEB-INF/jsp/playQueue.jsp:
```diff
@@ -426,6 +454,12 @@
             }
         }
 
+        if (internetRadioEnabled) {
```
This could be replaced with `$(".playqueue-actions").toggle(internetRadioEnabled);` I think

jvoisin commented on airsonic-main/src/main/webapp/WEB-INF/jsp/playQueue.jsp:
```diff
@@ -471,6 +518,11 @@
                 $("#albumUrl" + id).attr("href", song.albumUrl);
+                // Open external internet radio links in new windows
+                if (internetRadioEnabled) {
+                    $("#albumUrl" + id).attr("target", "_blank");
```
This can be replaced with the following:

```javascript
$( "#albumUrl" + id  ).attr({
  target: "_blank",
  rel: "noopener"
});
```

fxthomas replied:
Fixed that as well, that should be the last one ;)

fxthomas commented:
Rebased on `master`.

fxthomas commented:
I tested most of the radios on the front page of https://www.internet-radio.com, plus some of mine. It seems to work well if there are no redirects. I will add support for (a limited number of) HTTP redirects, since a lot of websites seem to use them.

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/service/InternetRadioService.java:
```diff
@@ -0,0 +1,226 @@
+    /**
+     * A list of cached source URLs for remote playlists.
+     */
+    private Map<Integer, List<InternetRadioSource>> cachedSources;
```
Can this be `final`?

fxthomas replied:
Yup.

jvoisin commented on airsonic-main/src/main/webapp/WEB-INF/jsp/playQueue.jsp:
```diff
@@ -99,6 +99,40 @@
+            $(this).each(function(ix, elt) {
+
+                var node, currentState;
+                if (elt.tagName.toLowerCase() === "a") currentState = true;
```
I think that you can use `elt.is("a")` instead.

fxthomas replied:
I tried, actually. Calling `$(elt)` results in a syntax error like `SyntaxError: '[object HTMLSpanElement]' is not a valid selector`.

I assume that's because we're using an old version of jQuery?

jvoisin replied:
We'll soon have the latest one via #930 :)

fxthomas commented:
So, I was not able to test it on Chromecast (but it "could" work as I set the remote URL correctly), and I am fairly certain that the Jukebox will not work at all because it only handles local files for now. I won't pursue this further, I think this should be ready for merging providing there are no other comments.

fxthomas commented:
I only did minimal work on the UI, so feel free if you feel there are things to improve.

Normal playlist:

![Screenshot from 2019-05-01 11-48-25](https://user-images.githubusercontent.com/613594/57012485-125a7600-6c07-11e9-94e8-80f77f9b33d5.png)

Radio UI:

![Screenshot from 2019-05-01 11-49-41](https://user-images.githubusercontent.com/613594/57012526-4170e780-6c07-11e9-885b-6598950e81c2.png)

pwepwe973 commented:
Hello, I wanted to know what file should you add to my.war file in the changes you made to correct this radio problem that does not start?
best regards pwepwe973

fxthomas commented:
If you can test [this test build](https://drive.google.com/open?id=19W7H0Qaa-p6HYQZll5v8QHOhCb-jj4Gh) (built on dd29d67) and tell me if it works for your internet radios, that'd be great!

pwepwe973 commented:
hello 
unable to download, google drive loads without ever launching the file download

pwepwe973 commented:
hello
I finally managed to download it with jdownloader
And I did the test and the reading of the radios is not done always when I click on the button play of the radio

fxthomas commented:
Ah, I guess Google does take a loooong time to give you the download link. Glad you could download and test it!

Yup, the navigator is going to read the radio, not another app. I implemented this by giving Airsonic the ability to download an external playlist (which is how most internet radios I've seen work, including the previous implementation in Subsonic), then parse the streaming links inside it and give them to Airsonic's media player.

_PS: Ton anglais marche bien, on comprend ce que tu veux dire! On est pas mal de Français par ici je crois :smile:_

pwepwe973 commented:
Hello 
thank you for my english or rather that of google translation: D

 When I say that it's the browser that reads the music, I mean that the player (.swf or html5 of airsonic "I do not know that type of player is lol")
  Ben he does not play music directly through the small player so we can not stop the radio or start another music because otherwise the music plays at the same time as the radio. which makes 2 music at the same time.

it will be nice to be able to stop the radio without updating the page or when you start a music that airsonic stops the radio and starts the music instead of playing the 2 at the same time.

here is my little problem :)

fxthomas commented:
Added unit tests, this should be ready to merge. @muff1nman, any chance this can make it for 10.3?

@pwepwe973 Can you confirm that you were running my dev build when you tested it? Airsonic version should be 10.3.0-SNAPSHOT in the "About" page. If not, you're still running the old broken internet radio implementation.

muff1nman commented on airsonic-main/src/main/java/org/airsonic/player/ajax/PlayQueueInfo.java:
```diff
@@ -33,17 +33,19 @@
     private final boolean repeatEnabled;
-    private final boolean radioEnabled;
+    private final boolean shuffleRadioEnabled;
+    private final boolean internetRadioEnabled;
```
I'm not hugely familiar with the radio feature as is so maybe I'm just naiive, but what is the difference between internetRadio and shuffleRadio?

fxthomas replied:
Shuffle Radio is what you get when you click on More > Shuffle Radio, kind of like a weak smart playlist with a few useful filters. Internet Radios are basically external playlists.

fxthomas replied:
(The two are mutually exclusive though, it might actually be a good idea to merge them in some kind of "mode" variable)

muff1nman replied:
Okay them being mutually exclusive is the part I was trying to understand and now I understand the need for two different variables.

fxthomas commented on airsonic-main/src/main/java/org/airsonic/player/service/InternetRadioService.java:
```diff
@@ -0,0 +1,226 @@
+        public PlaylistTooLarge(String message) {
+            super(message);
+        }
+    }
```
Right now they cannot be encountered from outside the class (the only methods that generate them are private). It's not a big change though, I can make them public.

pwepwe973 commented:
> Added unit tests, this should be ready to merge. @muff1nman, any chance this can make it for 10.3?
> 
> @pwepwe973 Can you confirm that you were running my dev build when you tested it? Airsonic version should be 10.3.0-SNAPSHOT in the "About" page. If not, you're still running the old broken internet radio implementation.


hello, yes when you asked me to test your test version that's what I did and the radio is not played in the player integrated airsonic.
![Capture d’écran de 2019-05-09 00-18-09](https://user-images.githubusercontent.com/37281042/57412455-0109fe80-71f1-11e9-9006-9c484d541ac4.png)
 But in the browser, basically impossible to stop reading of the radio with the button, it is necessary to update the page for that

fxthomas commented:
Uh, that's a bit annoying if it doesn't work somewhere... Let me try to figure out what went wrong:

* What browser are you using? IE, Firefox, Chrome? Another?
* What platform? Windows, Mac, Linux?
* Can you post relevant parts of the Airsonic logs? Something like `Retrieving playlist for internet radio xxxxx, got y sources.` should be there when you start playing an internet radio.
* Can you try to clear your cache then try again? (Shift-F5 or Shift-Ctrl-R)

pwepwe973 commented:
> Uh, that's a bit annoying if it doesn't work somewhere... Let me try to figure out what went wrong:
> 
>     * What browser are you using? IE, Firefox, Chrome? Another?
> 
>     * What platform? Windows, Mac, Linux?
> 
>     * Can you post relevant parts of the Airsonic logs? Something like `Retrieving playlist for internet radio xxxxx, got y sources.` should be there when you start playing an internet radio.
> 
>     * Can you try to clear your cache then try again? (Shift-F5 or Shift-Ctrl-R)

Hello 

- I use all the browser for the tests (firefox, chrome, and opera)
- I am under linux and sometimes windows.

- Airsonic is tale to him installed under openmediavault 4

I will attach a video capture to show the steps I'm doing

  https://youtu.be/XkamGtgYhvQ

I do not find info on the radio in the log even after clicking on the radio it adds no info in the log file.

 Log
--
2019-06-11 04:39:24.810  INFO --- o.a.p.service.PlaylistService            : Starting playlist import.
2019-06-11 04:39:24.811  INFO --- o.a.p.service.PlaylistService            : Completed playlist import.
2019-06-11 04:39:41.386  INFO --- o.a.p.c.StreamController                 : 127.0.0.1: Got HTTP range: 0-
2019-06-11 04:39:41.387  INFO ---  o.a.p.io.PlayQueueInputStream            : 127.0.0.1: prelvain  listening to T'es pas méchant/01) Teeyah - T'es pas méchant [T'es pas  méchant].mp3
2019-06-11 04:49:54.733  INFO --- o.a.p.c.StreamController                 : 127.0.0.1: Got HTTP range: 0-
2019-06-11 04:49:54.734  INFO ---  o.a.p.io.PlayQueueInputStream            : 127.0.0.1: prelvain  listening to Sale/01) Ya Levis - Sale [Sale].mp3
2019-06-11 04:50:02.047  INFO --- o.a.p.c.StreamController                 : 127.0.0.1: Got HTTP range: 0-
2019-06-11 04:50:02.048  INFO ---  o.a.p.io.PlayQueueInputStream            : 127.0.0.1: prelvain  listening to Liballa/01) Ya Levis - Liballa [Liballa].mp3
2019-06-11 04:52:37.767  INFO --- o.a.p.c.StreamController                 : 127.0.0.1: Got HTTP range: 0-
2019-06-11 04:52:37.770  INFO ---  o.a.p.io.PlayQueueInputStream            : 127.0.0.1: prelvain  listening to Liballa/01) Ya Levis - Liballa [Liballa].mp3
2019-06-11 04:53:11.124  INFO --- o.a.p.c.StreamController                 : 127.0.0.1: Got HTTP range: 0-
2019-06-11 04:53:11.124  INFO ---  o.a.p.io.PlayQueueInputStream            : 127.0.0.1: prelvain  listening to Liballa/01) Ya Levis - Liballa [Liballa].mp3

pwepwe973 commented:
as you can see in the videos, when I click on the radio the reading does not start in the player integrated with airsonic, but rather by the browser so there is no scrolling of the level of reading because it is the browser that read directly and when I click on a music in my library it mixes with the sound of the radio and they play both at the same time.

sory for my English i use google translate

fxthomas commented:
Rebased on master.

jvoisin commented:
What is left to do to get this merged?

fxthomas commented:
Nothing AFAIK.

It's been working well for the last month now. If there are remaining issues at least it's going to be better than the broken state of internet radios in master. Let's merge it! :)

jvoisin commented:
Time to invoke the @muff1nman to get his opinion on this, since it's a big PR :)

pwepwe973 commented:
> > Uh, that's a bit annoying if it doesn't work somewhere... Let me try to figure out what went wrong:
> > ```
> > * What browser are you using? IE, Firefox, Chrome? Another?
> > 
> > * What platform? Windows, Mac, Linux?
> > 
> > * Can you post relevant parts of the Airsonic logs? Something like `Retrieving playlist for internet radio xxxxx, got y sources.` should be there when you start playing an internet radio.
> > 
> > * Can you try to clear your cache then try again? (Shift-F5 or Shift-Ctrl-R)
> > ```
> 
> Hello
> 
>     * I use all the browser for the tests (firefox, chrome, and opera)
> 
>     * I am under linux and sometimes windows.
> 
>     * Airsonic is tale to him installed under openmediavault 4
> 
> 
> I will attach a video capture to show the steps I'm doing
> 
> https://youtu.be/XkamGtgYhvQ
> 
> I do not find info on the radio in the log even after clicking on the radio it adds no info in the log file.
> ##  Log
> 
> 2019-06-11 04:39:24.810 INFO --- o.a.p.service.PlaylistService : Starting playlist import.
> 2019-06-11 04:39:24.811 INFO --- o.a.p.service.PlaylistService : Completed playlist import.
> 2019-06-11 04:39:41.386 INFO --- o.a.p.c.StreamController : 127.0.0.1: Got HTTP range: 0-
> 2019-06-11 04:39:41.387 INFO --- o.a.p.io.PlayQueueInputStream : 127.0.0.1: prelvain listening to T'es pas méchant/01) Teeyah - T'es pas méchant [T'es pas méchant].mp3
> 2019-06-11 04:49:54.733 INFO --- o.a.p.c.StreamController : 127.0.0.1: Got HTTP range: 0-
> 2019-06-11 04:49:54.734 INFO --- o.a.p.io.PlayQueueInputStream : 127.0.0.1: prelvain listening to Sale/01) Ya Levis - Sale [Sale].mp3
> 2019-06-11 04:50:02.047 INFO --- o.a.p.c.StreamController : 127.0.0.1: Got HTTP range: 0-
> 2019-06-11 04:50:02.048 INFO --- o.a.p.io.PlayQueueInputStream : 127.0.0.1: prelvain listening to Liballa/01) Ya Levis - Liballa [Liballa].mp3
> 2019-06-11 04:52:37.767 INFO --- o.a.p.c.StreamController : 127.0.0.1: Got HTTP range: 0-
> 2019-06-11 04:52:37.770 INFO --- o.a.p.io.PlayQueueInputStream : 127.0.0.1: prelvain listening to Liballa/01) Ya Levis - Liballa [Liballa].mp3
> 2019-06-11 04:53:11.124 INFO --- o.a.p.c.StreamController : 127.0.0.1: Got HTTP range: 0-
> 2019-06-11 04:53:11.124 INFO --- o.a.p.io.PlayQueueInputStream : 127.0.0.1: prelvain listening to Liballa/01) Ya Levis - Liballa [Liballa].mp3

Hello 
did you see the videos posted in my post?
@fxthomas

fxthomas commented:
@pwepwe973: I did, and I could not reproduce the issue you have, either on Firefox or on Chromium, on 2 différent computers and on Windows/Linux...

pwepwe973 commented:
> @pwepwe973: I did, and I could not reproduce the issue you have, either on Firefox or on Chromium, on 2 différent computers and on Windows/Linux..



Hello 
so for you on your side it reads well in the reader here desous? :(
![Capture d’écran de 2019-05-09 00-18-09](https://user-images.githubusercontent.com/37281042/57412455-0109fe80-71f1-11e9-9006-9c484d541ac4.png)

fxthomas commented:
Yes, it's present on master and fixed on this branch for me. I'm wondering if something was maybe not reloaded properly, or if your server was maybe using the wrong .war file?

Can you try running the following command against the .war file you're using? It should show a class named `InternetRadioService` that was not present before this branch, like this (on the build I'm currently using):

```
root@yourserver ~ # zipinfo /var/lib/airsonic/airsonic.war | grep InternetRadioService.class 
-rw----     2.0 fat     8029 bl defN 19-Jul-02 23:00 WEB-INF/classes/org/airsonic/player/service/InternetRadioService.class
```

@muff1nman @jvoisin Can you guys test this and check that it works for you? If it does, we can merge this as a first attempt at fixing internet radios, then figure out what went wrong later!

muff1nman commented:
Not having luck so far in getting it to work. Notes:

1. Testing with https://www.internet-radio.com/station/wsjg/
2. NoScript block fired for the www.internet-radio.com domain - not sure why we're loading JS from there...
3. Putting in an exception for the domain, playback still did not fire. JS console log says: 

```
HTTP “Content-Type” of “text/html” is not supported. Load of media resource http://uk3.internet-radio.com:8021/ failed.
```

muff1nman commented:
Also: The test above was after a merge with master. I also just tried the PR as is and got a slightly different result:

```
http://uk3.internet-radio.com:8021/ playQueue.view:715:17
HTTP “Content-Type” of “text/html” is not supported. Load of media resource http://uk3.internet-radio.com:8021/ failed. playQueue.view
NotSupportedError: The media resource indicated by the src attribute or assigned media provider object was not suitable.
```

muff1nman commented:
Chrome: 
```
cast_framework.js:60 [Deprecation] document.registerElement is deprecated and will be removed in M73, around March 2019. Please use window.customElements.define instead. See https://www.chromestatus.com/features/4642138092470272 for more details.
(anonymous) @ cast_framework.js:60
playQueueCast.js:271 init success
playQueueCast.js:271 receiver list empty
playQueueCast.js:271 receiver found
playQueue.view:715 http://uk3.internet-radio.com:8021/
Cross-Origin Read Blocking (CORB) blocked cross-origin response http://uk3.internet-radio.com:8021/index.html?sid=1 with MIME type text/html. See https://www.chromestatus.com/feature/5629709824032768 for more details.
playQueue.view:1 Uncaught (in promise) DOMException
playQueueCast.js:271 receiver list empty
playQueueCast.js:271 receiver found
```

muff1nman commented:
Testing another source (`https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://uk6.internet-radio.com:8073/live.m3u&t=.pls`) seems to work okay. So maybe there is an issue with the other m3u source...

fxthomas commented:
Internet-radio.com has a few buggy sources that try to return HTML from the stream endpoint instead of actual audio data... Did you have an error in the UI when that happened?

muff1nman commented:
No error message.

fxthomas commented:
Okay, does not really surprise me. I'll try to work out a notification in case that happens in another PR!

jvoisin commented:
Wow, [CORB]( https://www.chromium.org/Home/chromium-security/corb-for-developers ) is live? Nice!

pwepwe973 commented:
Hello, 

here's what the command shows me...


**root@monserver:~# zipinfo /var/airsonic/airsonic.war | grep InternetRadioService.class 
-rw----     2.0 fat     8031 bl defN 19-May-04 10:14 WEB-INF/classes/org/airsonic/player/service/InternetRadioService.class**

fxthomas commented:
Uh. So I'm puzzled. You should obviously be getting the correct version (this class did not exist before the PR).

Let's try to dig up some more info (I'm on Firefox, Chrome should be very similar):

* Right-click on the ▶ symbol next to a radio, then choose "Inspect element". A window opens.
* In the window that opens, there should be a line with `<a class="radio-play" ...`, with a small "Event" button next to it. Click that button.

![image](https://user-images.githubusercontent.com/613594/60688203-06638c80-9eb4-11e9-8853-cebfec24b8f5.png)

* In the window that opens, one of the items should contain the following code:

![image](https://user-images.githubusercontent.com/613594/60688254-53476300-9eb4-11e9-960f-c1dfc96db721.png)

* In the "Console" tab, is there anything in red, or error messages? When you click on the "play" button to play a radio, what happens?

* Can you try to run `top.playQueue.onPlayInternetRadio(0, 0)` (replace the first "zero" with what's in the `data-id` parameter in the code shown in one of the steps above)? What happens? Does the radio play?

* Can you increase the log verbosity using the instructions in https://airsonic.github.io/docs/logging?

pwepwe973 commented:
there is all the info you ask me in the pages

![Capture du 2019-07-05 00-52-07](https://user-images.githubusercontent.com/37281042/60689656-0d909780-9ec0-11e9-9ffa-585e89304e95.png)

- In the console tab there is nothing displayed.

-When I click on play button radio plays, but not in the airsonic player at the bottom of the pages ... it's like in the video that I have enoyé above.

- Otherwise how to execute top.playQueue.onPlayInternetRadio (0, 0)

fxthomas commented:
@pwepwe973 Use the text box at the bottom of the "Console" tab. In your case try with `top.playQueue.onPlayInternetRadio (3, 0)` (because you have `data-id="3"` in your screenshot).

pwepwe973 commented:
TypeError: playQueueService.playInternetRadio is not a function

pwepwe973 commented:
Hello 
this radio works
http://player.playzoukantilles.com/playlist.php?name=listen.m3u

 but everyone else does not work :

http://www.zafemradio.net/tropical/tropical.m3u
http://cdn.nrjaudio.fm/adwz1/fr_an/55248/mp3_128.mp3
https://nrjguyane.ice.infomaniak.ch/nrjguyane-64.aac
http://195.154.240.161:8008/stream

fxthomas commented:
Damn, you're really out of luck with these...

* http://www.zafemradio.net/tropical/tropical.m3u and http://195.154.240.161:8008/stream use ShoutCAST, which uses a really old protocol and has issues with HTML5 audio (#1169)
* http://cdn.nrjaudio.fm/adwz1/fr_an/55248/mp3_128.mp3 and https://nrjguyane.ice.infomaniak.ch/nrjguyane-64.aac are direct audio links which are not yet supported (#1209)

ccardoso commented:
Hi @fxthomas can you please help me with a few questions: 
1 - How do you stream a user channel in airsonic, what configuration we need to have in airsonic a generate a link likes yours (http://IP:port/stream) ?

2- What is ShoutCAST and how we can configure them in airsonic if is possible ?

3- The DLNA/UPnP is used for streaming, how can configure this feature correctly in airsonic ?

I am a new in airsonic, that's why I put this dummy questions.

Many thanks

pwepwe973 commented:
> Hi @fxthomas can you please help me with a few questions:
> 1 - How do you stream a user channel in airsonic, what configuration we need to have in airsonic a generate a link likes yours (http://IP:port/stream) ?
> 
> 2- What is ShoutCAST and how we can configure them in airsonic if is possible ?
> 
> 3- The DLNA/UPnP is used for streaming, how can configure this feature correctly in airsonic ?
> 
> I am a new in airsonic, that's why I put this dummy questions.
> 
> Many thanks

Bonjour 

1- hello no airsonic just allows to read web radio not to broadcast webradio.

2 - SHOUTcast is the name of a protocol and a broadcasting server for webradio or webtv.
but as said in answer 1 aisonic just allows to read the webradios

3 I never managed to make DLNA work with airsonic that should not be in point because it works with everything at home emby, kodi, but airsonic nothing

ccardoso commented:
@pwepwe973 Thanks, but there is no way to stream a content in airsonic ?

pwepwe973 commented:
> @pwepwe973 Thanks, but there is no way to stream a content in airsonic ?

normally dlna is made to broadcast to your clients dlna of your network but I never managed to make work with airsonic but it must be an error that I made

jobenvil commented:
For many years I have been using minidlna, which works OK. Very simple.

I has no success with airsonic runing the DLNA feature as well. Lastly I compiled gerbera to use the DLNA feature, but somehow for simplicity and future maintaining work I abandoned the idea.

Until I switched to airsonic, I was using the payed version of subsonic, in which the DLNA feature was running pretty good. 

I will appreciate if someone got running the DLNA feaute and explain, how to achieve that.

pwepwe973 commented:
> Not having luck so far in getting it to work. Notes:
> 
> 1. Testing with https://www.internet-radio.com/station/wsjg/
> 2. NoScript block fired for the [www.internet-radio.com](http://www.internet-radio.com) domain - not sure why we're loading JS from there...
> 3. Putting in an exception for the domain, playback still did not fire. JS console log says:
> 
> ```
> HTTP “Content-Type” of “text/html” is not supported. Load of media resource http://uk3.internet-radio.com:8021/ failed.
> ```

j'ai testé les liens et cela ne fonction pas non plus. j,avais laissé tombé depuis ! a t'il une évolution au niveau de la lecture de tout type de webradio?
